### DISCUSSION TITLE
Access frontmatter after `compile` with `outputFormat: "function-body"`

### DISCUSSION THREAD

**Discussion by openscript** [mdx-js/mdx#2468] [Q&A] [answered]
Dear mdx-js community

I wonder if I can access the frontmatter when I `compile(page.default, { ...mdxOptions, outputFormat: "function-body" });` somehow without running?

I've read https://mdxjs.com/guides/frontmatter/ and loaded already `remarkPlugins: [remarkFrontmatter, remarkMdxFrontmatter]`.

Thank you for your help

**Comment by wooorm** [comment]
Yes, you should be able to access it when running when you use both those two plugins. Do a `console.log(await run(...))`, it’ll print the component at a `default` field, and will have other fields with your frontmatter!

**Reply by openscript** [reply]
Yes, I see that. Unfortunately I need the frontmatter already before running, so I can hand over the meta data to my framework for setting up the page.

**Reply by wooorm** [reply]
Then you don’t need ~~`remarkFrontmatter`,~~ `remarkMdxFrontmatter`, as those do the other part: exposing frontmatter when running.

If you don’t run, then yo can use any old frontmatter thing. As remco points out: `vfile-matter` is one by us that we often use!

**Reply by remcohaszing** [reply]
`remarkFrontmatter` is still needed to add frontmatter syntax support to the MDX content.

**Reply by openscript** [reply]
Thank you for your help! I didn't know about `vfile` before. What a nice and clean format!

I've seen that `compile` already creates a `vfile` and all the work with splitting off the frontmatter is already done by then. Wouldn't it be very handy to make frontmatter available in [`vfile.data`](https://github.com/vfile/vfile?tab=readme-ov-file#filedata)?

**Reply by wooorm** [reply, marked as answer]
That’s what `vfile-matter` indeed does!
`compile` indeed returns a vfile, with the compiled javascript as a result.
It also *accepts* a vfile, with the input markdown/mdx (with/without frontmatter).

You can put `vfile-matter` in a plugin, and pass that to `compile`. Then you get what you want too.
Or you can create a vfile first, call `vfile-matter` with that file, and pass that file to `compile`. That’s what I’d do.

**Comment by remcohaszing** [comment]
You can use [`vfile-matter`](https://github.com/vfile/vfile-matter) to access frontmatter data without running the code.
